### ISSUE TITLE
Any plans for RTL (right to left) support?

### ISSUE THREAD

**Issue by BrianJVarley** [JiHong88/SunEditor#450]
Hi @JiHong88 , the editor is impressive and covers nearly all of our requirements. But I don't see a config option for RTL  (right to left) text direction in the editor.

Is this something on the roadmap or already existing option? (I didn't see this in the docs). The rtl configuration is defined as..

> In a right-to-left, top-to-bottom script (commonly shortened to right to left or abbreviated RTL), writing starts from the right of the page and continues to the left. This can be contrasted against left-to-right writing systems, where writing starts from the left of the page and continues to the right.
> Arabic, Hebrew, Pashto, Persian, Urdu, and Sindhi are the most widespread RTL writing systems in modern times.

Example Toolbar button on other editor:

<img width="122" alt="rtl-editor" src="https://user-images.githubusercontent.com/6731968/88283984-7a9de780-cce4-11ea-88cb-e295fa4697a9.PNG">

Configuration Example:
![image](https://user-images.githubusercontent.com/6731968/88283933-57733800-cce4-11ea-8ab4-84afc1677ca1.png)

**Comment by JiHong88** [comment]
Hi @BrianJVarley 
Thank for your suggestion.
I'll add RTL option and button to next version.

**Comment by BrianJVarley** [comment]
Hey @JiHong88 Looking at the RTL support on different editors, the `RTL` button seems to be a common approach to flipping the editor content direction. But another thing to consider is order and direction of toolbar buttons. 


When I've tested in [this editor](https://ckeditor.com/ckeditor-4/demo/) you can see that certain toolbar icons are re-ordered and flipped backwards on language change to arabic:

<img width="959" alt="arabic-direction-rtl" src="https://user-images.githubusercontent.com/6731968/89054478-c4c04200-d350-11ea-8211-be017e946064.PNG">

**Comment by BrianJVarley** [comment]
@JiHong88 I see the commit above for RTL. Is this available to test on a public released version? https://github.com/JiHong88/SunEditor/commit/c43b5c511507025c152a1ef4fc5b50355bc29369

**Comment by JiHong88** [comment]
@BrianJVarley I will upload it to the test server.
Check and giving me feedback will help for develop.

**Comment by BrianJVarley** [comment]
@JiHong88 sure, How can I check this change on the test server?

**Comment by JiHong88** [comment]
@BrianJVarley I'll notify you at the this week after update.

**Comment by JiHong88** [comment]
Hi, @BrianJVarley 
I uploaded RTL options under development.
It is still being modified.
Modal window, the title at the bottom of the toolbar, and the table selector are being modified.

Please tell me know if there are any further modifications to be made.
Ex) Image modify box, etc...

http://suneditor.com/test/index.html

**Comment by BrianJVarley** [comment]
Hi @JiHong88 I tested the RTL feature above. The editor contents alignment and text typing direction look great for RTL.
But I notice some bugs with toolbar button direction, label direction and dropdown lists contents. When I compare it to this editor implementation of RTL - https://ckeditor.com/ckeditor-4/demo/(To test RTL change browser language to Arabic).

Here is some feedback on RTL feature that I think will improve it. I think if you can make the below changes it will make RTL feature it will be more complete:

- Toolbar Dropdown Menus (**Format, Size, Font, Paragraph Style, Text Style, Line Height, Align**): These toolbar button dropdown menus _contents_ should be LTR (left aligned), because contents are still in English and Numbers. The ▼ icon on dropdown should display on left side of button text:

<img width="590" alt="dropdown-menus-toolbar" src="https://user-images.githubusercontent.com/6731968/93593894-c7343500-f9ac-11ea-99eb-847520ccede1.PNG">

- Icon Direction: In RTL only _some icons_ should be reversed. I have highlighted in green icons that should be reversed, also highlighted in red icons should _not_ be reversed i.e, same as LTR mode:

<img width="792" alt="suneditor-icon-direction" src="https://user-images.githubusercontent.com/6731968/93595981-935b0e80-f9b0-11ea-9f99-96bbf8d9eb9f.png">


- Ordering of Icons: These icons should be ordered as, [Italic, Underline, Bold] the reverse of LTR mode. Also icons should not be flipped backwards. They should be same direction as LTR mode.

<img width="57" alt="bold-italic-underling-button" src="https://user-images.githubusercontent.com/6731968/93595409-9acde800-f9af-11ea-9e37-4c9317e66ee0.PNG">


- Undo Redo Buttons: These buttons should not be flipped. They are same in LTR mode as in RTL mode. See: https://material.io/design/usability/bidirectionality.html#mirroring-elements

![image](https://user-images.githubusercontent.com/6731968/93593115-593b3e00-f9ab-11ea-9e65-8046bcdc13e3.png)
 

- Toolbar Labels: Because toolbars labels are still in English, they should still read LTR. Currently, they are backwards.

<img width="590" alt="toolbar-label-backwards" src="https://user-images.githubusercontent.com/6731968/93592381-fbf2bd00-f9a9-11ea-87f9-258e73425b03.png">


- Image Modify Box:  URL input text should be LTR _not_ RTL, Submit button should be left side of modal, close (x) button should
be left side of modal. "Link" and "Image" tab text should be normal LTR as its English text. Tab should be on right side of modal.

<img width="590" alt="image-modal-rtl" src="https://user-images.githubusercontent.com/6731968/93592068-6c4d0e80-f9a9-11ea-97a2-37de38c7dbf0.png">

**Comment by JiHong88** [comment]
@BrianJVarley I updated to the [test page](http://suneditor.com/test/index.html).
And please comment on these components.

![1](https://user-images.githubusercontent.com/13191953/93778879-52d5ec00-fc61-11ea-93e5-3d3387a5a916.png)
![2](https://user-images.githubusercontent.com/13191953/93778884-54071900-fc61-11ea-81db-5cd2819c1a4b.png)
![3](https://user-images.githubusercontent.com/13191953/93779054-80229a00-fc61-11ea-9fe9-d12450558ca6.png)


Also the color picker is modified 
![4](https://user-images.githubusercontent.com/13191953/93778895-549faf80-fc61-11ea-93a9-8edc445d9f0e.png)

**Comment by BrianJVarley** [comment]
@JiHong88 The above changes look correct at first look and tollbar behaviour is correct after latest change. A good test to verify RTL is implemented correctly is to open [Google Docs](https://docs.google.com/document/u/0/) and change keyboard language to arabic. 

I do notice some minor bugs on label and toolbar buttons as follows:

1) Undo / Redo button: These buttons should be pointing same direction as in LTR mode. SO non change to icon, (left arrow = undo, right arrow - redo) See correct implementation here: 

![undo-redo-example](https://user-images.githubusercontent.com/6731968/93794997-be539580-fc30-11ea-8bf6-04b9e788a404.gif)


2) Indent / Out-dent button: The English part of the label here is backwards, should be LTR:

<img width="102" alt="indent-outdent-label" src="https://user-images.githubusercontent.com/6731968/93795105-db886400-fc30-11ea-8484-16afdcdb2dab.PNG">

**Comment by impulsgraw** [comment]
Hi @JiHong88 ,

Since some time I keep an eye on your wonderful wysiwyg library, it's really great!
I consider the RTL support as a must-have in modern content editors, it's amazing you've decided to implement it for the SunEditor. 
The suggestions from @BrianJVarley are quite precise 👍 .
During the implementation of RTL feature in my project I have found the following resources as comprehensively informative to refer to:

- https://translate.google.com/translate?hl=en&sl=ru&tl=en&u=https%3A%2F%2Fhabr.com%2Fru%2Fcompany%2F2gis%2Fblog%2F358148%2F
- https://translate.google.com/translate?hl=en&sl=ru&tl=en&u=https%3A%2F%2Fhabr.com%2Fru%2Fcompany%2Fabbyy%2Fblog%2F251579%2F
- https://www.w3.org/International/talks/1602-oman/ -- a nice post from W3
- https://steelkiwi.com/blog/right-to-left-rtl-website-development-design/

Several useful suggestions:

- Following things should always remain LTR (as they don't have RTL equivalents): phone numbers (including all signs), URLs, email addresses, date&time (unless it's in the Arabic/Hebrew traditional format)
- The default RTL locales for most languages (Arabic, Hebrew, Urdu) in the widely-used regular fonts (Times, Arial, Helvetica, etc) are quite poor and either just not intended for such purposes, or just appear unreadable for native speakers. Consider the example below -- such tooltip label has very small Arabic letters.
![image](https://user-images.githubusercontent.com/3176516/93924283-326b6780-fd1d-11ea-8765-36aec158bb7a.png)
  So I think it's better to reconsider the font-family. I have found the [**Noto Sans** family as the great alternative](https://www.google.com/get/noto/).
- BTW, on the screenshot with the tooltip above, the string slices are misaligned towards to each other: if in English you see something like "Shortcut description (Ctrl+F5)" then in Arabic you should see "(F5+Ctrl) Shortcut description" -- the part in the round braces moves from the string ending into the beginning (and vice-versa).

Don't hesitate to reach out in case of questions!
I'll try to have some more walkthrough meantime,

Thanks,
Andrey.

**Comment by JiHong88** [comment]
@BrianJVarley @impulsgraw 
I updated RTL mode of the test page. http://suneditor.com/test/index.html

The creation of all controllers and tables has also been modified with RTL.
And the list of align buttons has been modified so that the right sort is at the top by default.

At the bottom of the test page is a list of buttons used when creating the editor.
I created a button module by reversing the order of the button list in RTL mode, but I am not sure if it is the correct operation.

And..  I'm confused if the Math plugin should also be RTL.
